File: streaming-server/src/main/java/io/github/hs96wings/streaming_server/comment/service/CommentService.java
package io.github.hs96wings.streaming_server.comment.service;

import io.github.hs96wings.streaming_server.comment.domain.Comment;
import io.github.hs96wings.streaming_server.comment.dto.CommentResDto;
import io.github.hs96wings.streaming_server.comment.dto.CommentSaveReqDto;
import io.github.hs96wings.streaming_server.comment.repository.CommentRepository;
import io.github.hs96wings.streaming_server.member.domain.Member;
import io.github.hs96wings.streaming_server.member.repository.MemberRepository;
import io.github.hs96wings.streaming_server.video.domain.Video;
import io.github.hs96wings.streaming_server.video.repository.VideoRepository;
import jakarta.persistence.EntityNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Transactional
public class CommentService {
    private final CommentRepository commentRepository;
    private final VideoRepository videoRepository;
    private final MemberRepository memberRepository;

    public CommentService(CommentRepository commentRepository, VideoRepository videoRepository, MemberRepository memberRepository) {
        this.commentRepository = commentRepository;
        this.videoRepository = videoRepository;
        this.memberRepository = memberRepository;
    }

    public CommentResDto addComment(CommentSaveReqDto commentSaveReqDto, String username) {
        Video video = videoRepository.findById(commentSaveReqDto.getVideoId())
                .orElseThrow(() -> new EntityNotFoundException("영상이 없습니다"));
        Member member = memberRepository.findByUserid(username)
                .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 유저입니다"));

        Comment comment = Comment.builder()
                .video(video)
                .author(member)
                .content(commentSaveReqDto.getContent())
                .build();

        commentRepository.save(comment);

        return new CommentResDto(comment);
    }

    @Transactional(readOnly = true)
    public List<CommentResDto> getComments(Long videoId) {
        return commentRepository.findByVideoIdOrderByCreatedAtAsc(videoId)
                .stream()
                .map(CommentResDto::new)
                .toList();
    }

    public void deleteComment(Long commentId, String username) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 댓글입니다"));

        if (!comment.getAuthor().getUserid().equals(username)) {
            throw new org.springframework.security.access.AccessDeniedException("작성자만 삭제할 수 있습니다");
        }

        commentRepository.delete(comment);
    }
}

File: streaming-server/src/main/java/io/github/hs96wings/streaming_server/comment/service/CommentService.kt
package io.github.hs96wings.streaming_server.comment.service

import io.github.hs96wings.streaming_server.comment.domain.Comment
import io.github.hs96wings.streaming_server.comment.dto.CommentResDto
import io.github.hs96wings.streaming_server.comment.dto.CommentSaveReqDto
import io.github.hs96wings.streaming_server.comment.repository.CommentRepository
import io.github.hs96wings.streaming_server.member.domain.Member
import io.github.hs96wings.streaming_server.member.repository.MemberRepository
import io.github.hs96wings.streaming_server.video.domain.Video
import io.github.hs96wings.streaming_server.video.repository.VideoRepository
import jakarta.persistence.EntityNotFoundException
import org.springframework.security.access.AccessDeniedException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class CommentService(
    private val commentRepository: CommentRepository,
    private val videoRepository: VideoRepository,
    private val memberRepository: MemberRepository
) {
    fun addComment(commentSaveReqDto: CommentSaveReqDto, userid: String): CommentResDto {
        val video = findVideoById(commentSaveReqDto.videoId)
        val member = findMemberByUserId(userid)

        val comment = Comment(
            commentSaveReqDto.content,
            video,
            member
        )

        val savedComment: Comment = commentRepository.save(comment)
        return CommentResDto.from(savedComment)
    }

    fun getComments(videoId: Long): List<CommentResDto> {
        return commentRepository.findByVideoIdOrderByCreatedAtAsc(videoId).map(CommentResDto::from)
    }

    @Transactional
    fun deleteComment(commentId: Long, username: String) {
        val comment: Comment = findCommentById(commentId)

        if (comment.author.userid != username) {
            throw AccessDeniedException("작성자만 삭제할 수 있습니다")
        }

        commentRepository.delete(comment)
    }

    private fun findVideoById(videoId: Long): Video {
        return videoRepository.findById(videoId)
            .orElse(null) ?: throw IllegalArgumentException("해당 영상이 존재하지 않습니다. id=$videoId")
    }

    private fun findMemberByUserId(userid: String): Member {
        return memberRepository.findByUserid(userid)
            .orElse(null) ?: throw EntityNotFoundException("존재하지 않는 유저입니다")
    }

    private fun findCommentById(commentId: Long): Comment {
        return commentRepository.findById(commentId)
            .orElse(null) ?: throw EntityNotFoundException("존재하지 않는 댓글입니다")
    }
}
